### ISSUE TITLE
Show corresponding remote branch of local branch

### ISSUE THREAD

**Issue by Saafo** [jesseduffield/lazygit#2751]
**Is your feature request related to a problem? Please describe.**
I have a large repo and the remote branch is by thousands, but the local branch is less than 10 usually. I want to check the latest commit of the remote branch corresponding to my local branch **before I fast-forward my local branch**

**Describe the solution you'd like**
Show the latest commits both in remote and local in git log and branch commits list like that in Tower.

**Describe alternatives you've considered**
Jump to corresponding remote branch from local branch and check the commit list. At least can avoid search branch in thousands of remote branches each time.

**Comment by stefanhaller** [comment]
Yes, I want this too. My current solution for this is described in the third bullet point of [this comment](https://github.com/jesseduffield/lazygit/issues/2659#issuecomment-1556618462); it's a hacky branch that I didn't get around to cleaning up yet in order to make a PR for it. Also, I'm not really sure that the UX I chose is the best one, because, while it's convenient for comparing a branch against its upstream, sometimes you also want to compare arbitrary branches in the same way (e.g. a branch against master), and my approach doesn't allow that.

I'd be interested in how Tower shows the information, could you post a screenshot?

**Comment by Saafo** [comment]
> I'd be interested in how Tower shows the information, could you post a screenshot?

Sure:
<img width="611" alt="image" src="https://github.com/jesseduffield/lazygit/assets/46736350/85afa691-9948-487b-8274-10a637c6a2a7">

**Comment by stefanhaller** [comment]
Ah, interesting. So when hitting enter on a branch in the branches panel, it would simply do the `git log` on both the branch and its upstream. Sounds like a relatively easy change, I'll play around with this a bit.

One thing to consider is that the feature would only work well if `git.log.showGraph: always` is configured, otherwise the log would be very confusing. Also, I'm not sure the current graph implementation even supports this case, but we'll see.

**Comment by Saafo** [comment]
Show commits only in `git log` area may be not enough, I expected to show in commit list, so we can check changes per commit.

**Comment by mark2185** [comment]
The easiest approach might be to add a mapping for showing upstream changes so that `Enter` would show the local branch, while e.g. `G` would show the same `commits view`, but of the upstream branch, the same what would happen if the user manually searched for the branch in one of the `Remotes` and pressing `Enter` there.

**Comment by stefanhaller** [comment]
> Show commits only in `git log` area may be not enough, I expected to show in commit list, so we can check changes per commit.

Yes, that's what I meant. I talked about hitting enter on a branch. That's also why the `git.log.showGraph: always` setting is relevant here.

Actually I'm not sure it will really work as well in lazygit as it does in Tower, because we don't have the room to show branch names (except in the expanded view).

**Comment by stefanhaller** [comment]
> The easiest approach might be to add a mapping for showing upstream changes so that `Enter` would show the local branch, while e.g. `G` would show the same `commits view`, but of the upstream branch, the same what would happen if the user manually searched for the branch in one of the `Remotes` and pressing `Enter` there.

Did you see my proposal [here](https://github.com/jesseduffield/lazygit/issues/2751#issuecomment-1612471661)? That's kind of what I'm doing in my hack (I chose to bind it to alt-enter), except that I show the equivalent of what you would get when you type `git log --oneline --left-right branch...branch@{u}`. I find this more useful than seeing just the remote branch log and having to go back and forth to see the differences.

**Comment by mark2185** [comment]
> Did you see my proposal https://github.com/jesseduffield/lazygit/issues/2751#issuecomment-1612471661?

I actually did not, I saw the PR but I never even heard of triple dot log/diff before so I haven't delved too deep into the discussion, but that _hack_ does seem exactly what would solve the issue, yeah :)

**Comment by stefanhaller** [comment]
Here's a PR that implements this, feedback welcome: #2871.

**Comment by Saafo** [comment]
I think this issue is solved by #2871, now I use the divergence feature much often, thanks @stefanhaller !
